### ISSUE TITLE
Differing dependencies of matplotlib 2.0 with Python 3 and Python 2

### ISSUE THREAD

**Issue by HolgerPeters** [matplotlib/matplotlib#7893]
It seems matplotlib now uses the `functools32` library on Python 2, which is not needed on Python 3, because `functools32` is a backport of Python 3's built in functionality.

This is an issue to us, because we like to compile a list of pinned requirements for Python projects that define the testing environment. Compiling requirements with [pip-tools](https://github.com/nvie/pip-tools/) no yields different requirement lists on Python 3 and Python 2. So we would need to now keep track of two lists of requirements, one for Python 3 and one for Python 2 if we want to test with both environments.

Don't really know how this could be optimally addressed, but maybe you could incorporate the `lru_cache` function directly into your codebase and remove the dependency (since I think `lru_cache` is the only function used from functools32). Or I could report this to the upstream `functools32` because I think one could write `functools32` in a backward compatible way so we wouldn't have the issue that we cannot install `functools32` on Python 3.

**Comment by anntzer** [comment]
Would #7871 suit you?  (Or you could report an issue upstream.)

**Comment by tacaswell** [comment]
I am :-1: on vendoring parts of `functools32`.

```diff
diff --git a/functools32/__init__.py b/functools32/__init__.py
index 837f7fb..80151f7 100644
--- a/functools32/__init__.py
+++ b/functools32/__init__.py
@@ -1 +1,4 @@
-from .functools32 import *
+try:
+    from functools import *
+except ImportError:
+    from .functools32 import *
diff --git a/setup.py b/setup.py
index a04c5d9..40752f1 100755
--- a/setup.py
+++ b/setup.py
@@ -5,9 +5,6 @@ from distutils.core import setup
 
 
 def main():
-    if not (3,) > sys.version_info >= (2, 7):
-        sys.stderr.write('This backport is for Python 2.7 only.\n')
-        sys.exit(1)
 
     setup(
       name='functools32',
diff --git a/test_functools32.py b/test_functools32.py
index b709f37..e7c800a 100644
--- a/test_functools32.py
+++ b/test_functools32.py
@@ -6,7 +6,7 @@ import test_support27 as support
 from weakref import proxy
 import pickle
 from random import choice
-from functools32.functools32 import OrderedDict
+from collections import OrderedDict
 
 @staticmethod
 def PythonPartial(func, *args, **keywords):
diff --git a/test_support27.py b/test_support27.py
index 8b85f1c..4a4bfeb 100644
--- a/test_support27.py
+++ b/test_support27.py
@@ -15,7 +15,7 @@ import shutil
 import warnings
 import unittest
 import importlib
-import UserDict
+from collections import MutableMapping as DictMixin
 import re
 import time
 try:
@@ -77,7 +77,7 @@ def import_module(name, deprecated=False):
     with _ignore_deprecated_imports(deprecated):
         try:
             return importlib.import_module(name)
-        except ImportError, msg:
+        except ImportError as msg:
             raise unittest.SkipTest(str(msg))
 
 
@@ -188,7 +188,7 @@ def unlink(filename):
 def rmtree(path):
     try:
         shutil.rmtree(path)
-    except OSError, e:
+    except OSError as e:
         # Unix returns ENOENT, Windows returns ESRCH.
         if e.errno not in (errno.ENOENT, errno.ESRCH):
             raise
@@ -382,10 +382,10 @@ else:
             except UnicodeEncodeError:
                 pass
             else:
-                print \
+                print( \
                 'WARNING: The filename %r CAN be encoded by the filesystem.  ' \
                 'Unicode filename tests may not be effective' \
-                % TESTFN_UNENCODABLE
+                % TESTFN_UNENCODABLE)
 
 
 # Disambiguate TESTFN for parallel testing, while letting it remain a valid
@@ -495,7 +495,7 @@ def open_urlresource(url, check=None):
     # Verify the requirement before downloading the file
     requires('urlfetch')
 
-    print >> get_original_stdout(), '\tfetching %s ...' % url
+    print ('\tfetching %s ...' % url, file=get_original_stdout())
     f = urllib2.urlopen(url, timeout=15)
     try:
         with open(fn, "wb") as out:
@@ -653,7 +653,7 @@ class CleanImport(object):
         sys.modules.update(self.original_modules)
 
 
-class EnvironmentVarGuard(UserDict.DictMixin):
+class EnvironmentVarGuard(DictMixin):
 
     """Class to help protect the environment variable properly.  Can be used as
     a context manager."""
@@ -678,8 +678,11 @@ class EnvironmentVarGuard(UserDict.DictMixin):
         if envvar in self._environ:
             del self._environ[envvar]
 
-    def keys(self):
-        return self._environ.keys()
+    def __len__(self):
+        return len(self._environ.keys())
+
+    def __iter__(self):
+        return iter(self._environ)
 
     def set(self, envvar, value):
         self[envvar] = value
@@ -1118,7 +1121,7 @@ def run_doctest(module, verbosity=None):
     finally:
         sys.stdout = save_stdout
     if verbose:
-        print 'doctest (%s) ... %d tests with zero failures' % (module.__name__, t)
+        print('doctest (%s) ... %d tests with zero failures' % (module.__name__, t))
     return f, t
 
 #=======================================================================
```

Is a patch to `functools32` that makes its tests (mostly) pass on 3.6 (1 test that touches internals fails, 2 tests raise the wrong exception (`TypeError` vs `AttributeError`).

**Comment by HolgerPeters** [comment]
@anntzer I think #7871 would fix my issue

So would probably fixing the supstream package.

**Comment by anntzer** [comment]
Closed by #7871.
